python/tubix20_reframer.py

```diff
@@ -0,0 +1,176 @@
+     Amateur_Radio/TechnoSat_Telemetry_Format.ods>
+[2]: <https://destevez.net/2016/09/some-notes-on-beesat-and-mobitex-nx/>
+"""
+from gnuradio import gr
+import pmt
+
+from .check_eseo_crc import crc16_ccitt_zero as crc16_ccitt_zero
+
+
+class tubix20_reframer(gr.basic_block):
+    """
+    Input:
+        PDUs with decoded mobitex data blocks.
+        Each PDU must have a 'block id' tag.
+        The first PDU must have the following additional tags:
+            - 'num_blocks' tag: Number of blocks for this frame
+            - 'frame_header' tag: Header of this frame (typically 11 bytes)
+                - 2 control bytes
+                - 1 control bytes FEC
+                - 6 callsign bytes
+                - 2 callsign CRC bytes
+    Output:
+        PDUs with Tubix20 "Master Frames"
+        - master frame header (typically 10 bytes)
+            - 2 control bytes
+            - 6 callsign bytes
+            - 2 callsign CRC bytes
+        - master frame body
+            - N x 18 bytes
+        - master frame appendix
+            - 1 byte "ground station quality byte"
+            - 4 bytes "error marker"
+            - 1 byte "ground station TNC temperature"
+    """
+    bytemap = {
+        'control_bytes': slice(0, 2),
+        'control_bytes_crc': slice(2, 3),
+        'callsign': slice(3, 9),
+        'callsign_crc': slice(9, 11),
+    }
+
+    def __init__(self):
+        gr.basic_block.__init__(
+            self,
+            name='tubix20_reframer',
+            in_sig=[],
+            out_sig=[])
+
+        self.message_port_register_in(pmt.intern('in'))
+        self.set_msg_handler(pmt.intern('in'), self.handle_msg)
+        self.message_port_register_out(pmt.intern('out'))
+        self.reset()
+
+    def reset(self):
+        self.num_blocks_expected = 0
+        self.frame_header = None
+
+        self.block0_meta = None
+        self.blocks = {}
+        self.tags = {}
+
+        self.timestamp = None
+
+    def handle_msg(self, msg_pmt):
+        meta = pmt.car(msg_pmt)
+        packet = pmt.u8vector_elements(pmt.cdr(msg_pmt))
+        tags = pmt.to_python(meta)
+
+        # Reset state on block 0
+        if tags['block_id'] == 0:
+            if len(self.blocks) != 0:
+                print('ERROR: Received new block 0 while there are still '
+                      'unhandled blocks in the buffer.')
+                # Discard partial frame
+                self.reset()
+
+            self.num_blocks_expected = tags['num_blocks']
+            self.frame_header = tags['frame_header']
+
+            meta = pmt.dict_delete(meta, pmt.intern('block_id'))
+            meta = pmt.dict_delete(meta, pmt.intern('frame_header'))
+            meta = pmt.dict_delete(meta, pmt.intern('crc_valid'))
+            meta = pmt.dict_delete(meta, pmt.intern('corrected_errors'))
+            meta = pmt.dict_delete(meta, pmt.intern('uncorrected_errors'))
+            self.block0_meta = meta
```

Comment:
Maybe create a new empty dict and copy the data that needs to remain?

Reply:
> My goal was to transparently forward all unknown tags from the first data block to the output frame.

That is a good idea. I didn't have that in mind when I wrote the comment.

Your rewritten code looks good.